shumai/module/transformer.ts

```diff
@@ -0,0 +1,128 @@
+    return tensor.mul(this.scale_factor)
+  }
+
+  forward(queries: Tensor, keys: Tensor, values: Tensor): Tensor {
+    // shape [..., tokens, dim]
+    const shape = queries.shape
+    for (let i = 0; i < shape.length; i++) {
+      if (shape[i] !== keys.shape[i] || shape[i] !== values.shape[i]) {
+        throw new Error(
+          `Input tensors should have the same shape: queries shape ${shape}, keys shape ${keys.shape}, values shape ${values.shape}`
+        )
+      }
+    }
+
+    const dim = shape[shape.length - 1]
+    if (dim !== this.dim) {
+      throw new Error(
+        `Last axis of input tensors (shape ${shape}) must match module dimensions (${this.dim})`
+      )
+    }
+
+    let output = queries.matmul(keys.T()) // shape [..., tokens, tokens]
+    output = this.scale(output).softmax(-1)
+    output = output.matmul(values) // shape [..., tokens, dim]
+    return output
+  }
+}
+
+export class TransformerMultiheadAttention extends Module {
+  private dim: number
+  private heads: number
+  private attn_dim: number
+  private query_embed: sm.module.Linear
+  private key_embed: sm.module.Linear
+  private value_embed: sm.module.Linear
+  private attention: TransformerDotProductAttention
+  private concat_embed: sm.module.Linear
+
+  constructor(dim: number, heads: number, attn_dim?: number) {
+    super()
+
+    if (dim % heads !== 0) {
+      throw new Error(
+        `Model dimensions must be divisible by the number of heads: ${dim} not divisible by ${heads}`
+      )
+    }
+
+    this.dim = dim
+    this.heads = heads
+    if (attn_dim === undefined) {
+      this.attn_dim = dim / heads
+    } else {
+      this.attn_dim = attn_dim
+    }
+    this.query_embed = new sm.module.Linear(dim, this.attn_dim * heads)
+    this.key_embed = new sm.module.Linear(dim, this.attn_dim * heads)
+    this.value_embed = new sm.module.Linear(dim, this.attn_dim * heads)
+    this.attention = new TransformerDotProductAttention(this.attn_dim)
+    this.concat_embed = new sm.module.Linear(this.attn_dim * heads, dim)
+  }
+
+  forward(queries: Tensor, keys: Tensor, values: Tensor): Tensor {
+    // shape [..., tokens, dim]
+    const shape = queries.shape
+    for (let i = 0; i < shape.length; i++) {
+      if (shape[i] !== keys.shape[i] || shape[i] !== values.shape[i]) {
+        throw new Error(
+          `Input tensors should have the same shape: queries shape ${shape}, keys shape ${keys.shape}, values shape ${values.shape}`
+        )
+      }
+    }
+
+    const dim = shape[shape.length - 1]
+    if (dim !== this.dim) {
+      throw new Error(
+        `Last axis of input tensors (shape ${shape}) must match module dimensions (${this.dim})`
+      )
+    }
+
+    const reshape = [...shape] // [..., tokens, dim]
+    reshape[reshape.length - 1] = this.heads
+    reshape.push(this.attn_dim) // [..., tokens, heads, attn_dim]
+
+    // Swap 2nd and 3rd last axes
+    const transpose = reshape.map((x, i) => i)
```

Comment:
not sure its too much cleaner, but you can do `Array.from(sm.util.range(reshape.length))`